### ISSUE TITLE
Get locations of cells with invalid types

### ISSUE THREAD

**Issue by noctuid** [ToucanToco/fastexcel#308]
On 0.12.0, I'm not seeing an error if there is a row with e.g."foo" for a column that is set to "date" in "dtypes".  That cell just is read as null.  A string in a float column is also just converted to null.  Is this expected?  It would be great there was some way to get the locations of all failed cells, so that we could give the user back a file with bad cells highlighted.

**Comment by lukapeschke** [comment]
It is the expected behavior, yes. The rationale is that if the dtype is specified, the user knows what kind of data is contained in the column. 

I reckon it'd be nice to at least have a log in case the cell is of an unexpected type, but that would require work on calamine's side. We could probably add some methods to the `DataType` trait which would return `Result<Option<_>>`, with an error in case the cell is not empty, but isn't of the expected dtype either.

However we'd need to think a bit about this, since we don't want to affect performance (maybe through a new `warn_on_unexpected_dtype` parameter ?).

@PrettyWood what are your thoughts on this ?

**Comment by lukapeschke** [comment]
Yes, calamine kinda silently converts unexpected data to nulls: it provides `is_T() -> bool` methods for every excel datatype, along with `get_T() -> Option<T>` . The `get_T` methods return `None` if they are called on an unexpected variant (`get_datetime` on an int for example).  You're supposed to call `is_T` before using `get_T`.

But for columns where the dtype is provided, fastexcel skips the inference step, since it requires iterating over a lot of values. This means, we directly use `get_T`. A solution to this would be to add `try_get_T` methods to calamine, which would allow to get an error in case the data is of an unexpected type. 

But we'd also need to design an API around this in fastexcel: do we want to load the data with an attribute indicating that the cell contains invalid data ? Do we want to provide a parameter indicating what to do with errors ? This could be `'ignore' | 'store' | 'raise'` . 

I'm afraid there's no quick and easy solution for this one

**Comment by noctuid** [comment]
How do you imagine "store" would work?

**Comment by lukapeschke** [comment]
@noctuid As a user, I'd probably want an object separate from the data, which would contain the locations and errors for invalid cells

**Comment by noctuid** [comment]
I will probably get some dedicated time for my job to look into this in a couple of weeks. @lukapeschke  @PrettyWood If either of you have any other feedback/suggestions for implementing this, that would be great.

I have not tried setting up the project yet.  Since this will also require a change to calamine, is it easily possible to use a local version of calamine with fastexcel?

**Comment by lukapeschke** [comment]
@noctuid Thanks for your interest! Yes, you can use a local calamine version, by changing calamine to a local dependency in fastexcel's `Cargo.toml`

Regarding the implem, you can discuss with them what kind of API they'd like for error feedback, and if they'd be fine with the `try_get_T` API.

On fastexcel's side, I think that for a first step you should create a separate `to_arrow_with_error` method on the `ExcelSheet` type, with separate data loading functions, so we can check the performance impact of the `try_` versions of calamine accessors. If that method could return a `(PyAny, DataErrors)` tuple it would be great. Feel free to ask for help/comments if needed, I'll do my best to help 🙂

**Comment by noctuid** [comment]
Would DataErrors just be an array of Cell positions?

**Comment by lukapeschke** [comment]
A wrapper struct around an array is enough for now, yes. It will allow to add some helper methods to it

**Comment by noctuid** [comment]
Can you let me know if I'm understanding what you had in mind correctly?

- Add try_get functions in calamine that return `Result`s
- Add new versions of all functions in array_impls that additionally return the cells locations with errors
- Everything up from that returns the error locations as well: a new RecordBatch try_from equivalent that also returns all the error locations, a new `to_arrow_with_error`, and then what? Alter `build_sheet` to optionally use `to_arrow_with_error` and add a `new load_sheet_with_errors`?

> so we can check the performance impact of the try_ versions of calamine accessors

I will start with separate functions for testing, but do you anticipate this having a noticeable performance impact?  IIUC the only difference from the `get` version is an extra check that the value is None (None -> None, valid -> value, otherwise return Err instead of returning None like the original `get` functions do).

If the performance is fine, would the code just always use the `try_get` versions?

**Comment by noctuid** [comment]
Taking a closer look at the code now:

It looks like fastexcel does use the `get_T` functions (mainly for `create_string_array`) but mostly uses the `as_T` functions; in some cases these do use the `is_T` functions (e.g. `as_datetime`).  Please correct me if I'm wrong, but I don't think any changes to calamine are actually required.  Instead of adding alternate versions of all the `as_T` and `get_T` functions in calamine, it might make more sense to just check `is_empty` first in the `create_T_array` functions (and return None).  Otherwise if whatever calamine functions are used result in None, that means parsing failed, and we can store that row/col as a failure.  Would this implementation be acceptable, and do you think we need separate versions of the `create_T_array` functions (e.g. `create_T_array_with_error`), or can I just modify the originals and just always collect the parsing failure locations?

**Comment by lukapeschke** [comment]
@noctuid If possible, we'd rather avoid extra checks in fastexcel to keep things fast, so I'd prefer separate `create_T_array_with_error` functions to collect failures. Regarding the `try_get_T` versions in calamine, it can be done without them, but they'd allow for better error messages, such as "expected X1 or X2, but got Y"

**Comment by noctuid** [comment]
@lukapeschke It will probably be simpler to have them as separate functions, so I'll do it that way, though I think there are likely much bigger factors in performance.  Fastexcel/calamine performance gets significantly worse at some point, e.g. ~9.2s for reading a 55MB 600k excel file to ~157s for a 160MB 884k row file we have.  I actually didn't see any difference in read time for those files before/after my error location change.

> but they'd allow for better error messages, such as "expected X1 or X2, but got Y"

I can give a message like `Could not parse SharedString("bad int") as int` without any calamine changes or `Could not parse "bad int" as int` by just updating DataRef to implement Display.  What other information would you want in an error message?

I initially started by modifying calamine, but it seemed like unnecessary code duplication of all the `as_T` and `get_T` functions without a clear benefit.

What is the minimal PR you would accept?  Would it be enough initially for me to just add a `to_arrow_with_errors` method that returned the suggested tuple where DataErrors is now a vector of a DataError struct with "row" and "column"?  "error" could maybe be added later if it is not possible to give the error message you'd like without calamine changes?  It will be more difficult to convince our product manager to give me time to clean up what I have and contribute it back to fastexcel if it will require larger PRs to multiple repos from the start.

**Comment by lukapeschke** [comment]
@noctuid Yes, just having `to_arrow_with_errors` returning the suggested tuple is fine with me 🙂 You can do it without modifying calamine for now

Could you please just make `DataErrors` a wrapper struct around a `Vec` ? You can make `errors` a public property, but this would allow us to add features to the `DataErrors` type later on without breaking the API.

Thanks for taking care of this!

**Comment by noctuid** [comment]
Thanks, can you provide the struct definition you want just so I'm completely clear on what you mean?

**Comment by lukapeschke** [comment]
For now, something like this would be enough

```rust
pub struct CellError {
   pub position: (usize, usize),
   pub detail: String,
   // More stuff here if you want to, such as the original error
}

pub struct CellErrors {
  errors: Vec<CellError>
}
```

**Comment by noctuid** [comment]
@lukapeschke Thanks again for your help. I opened a PR with the initial changes for the new method.

**Comment by lukapeschke** [comment]
Great, thanks a lot! I'll try to review it this afternoon
